### DISCUSSION TITLE
Can we support public endpoint for the CodeSpace port forwarding URL?

### DISCUSSION THREAD

**Discussion by dooriya** [community/community#3637] [Codespaces]
For those backend APIs hosting in CodeSpaces environment, if we access the API by using the port forward endpoint (https://sample-project-7mfc-5000.githubpreview.dev/), it would **redirect to the GitHub authentication page**. Unlike a front-end app, this login redirect will be unable to process for a backend API call, and a preflight request for this backend API all will also failed due to the CORS policy: *redirect is not allow for a preflight request*.

![image](https://user-images.githubusercontent.com/10163840/112798140-0ef3f180-909f-11eb-960e-43e800c0deeb.png)

![image](https://user-images.githubusercontent.com/10163840/112798649-c983f400-909f-11eb-88b6-5ec3df221855.png)

Similar requirement can be also found at GitHub Community: https://github.community/t/expose-public-port/133598

**Comment by synsh** [comment]
There is a new privacy option. You can make it public, and auth redirection won't happen.


![Screenshot from 2021-03-31 09-39-11](https://user-images.githubusercontent.com/14250910/113089633-ae32f900-9205-11eb-99e5-9778b58c0f35.png)

**Reply by 2percentsilk** [reply]
Indeed! This feature just went out this week 😄 
We'll be updating our docs to make this more public soon!

Give it a try and let us know if you have any issues!

**Reply by dooriya** [reply]
Thank you @sahasayan and @2percentsilk for the update!
I just have a quick try for this feature, but I still got auth redirection after I make the endpoint public :(.

![image](https://user-images.githubusercontent.com/10163840/113370657-ffbebd80-9396-11eb-8210-516ddb10942a.png)

![image](https://user-images.githubusercontent.com/10163840/113370730-3268b600-9397-11eb-91a6-37400085bae7.png)

BTW, instead **manually enable** the endpoint public from the VS Code UI, is there any way to set the forward endpoint (for some port) to be public by default? And we don't want our customer to manually "Make Public" each time they start the app in CodeSpace.

**Reply by NovqiGarrix** [reply]
BTW, instead manually enable the endpoint public from the VS Code UI, is there any way to set the forward endpoint (for some port) to be public by default? And we don't want our customer to manually "Make Public" each time they start the app in CodeSpace.

**_Reply:_**
Hy, there. I also ran into this issue. Every time my Node.js server restarts Codespace forward the port with private visibility. 

One way to solve this is by forwarding your port manually and setting the visibility to public. 
So, even if Node.js restarts your server, the port will always be public.

**Comment by dooriya** [comment]
Hi @2percentsilk and @sahasayan,

May I know the current status of the public endpoint support for CodeSpace port forwarding?
I just retry today, after I click "Make Public", the endpoint still redirect to auth. What's more, the the port-forwarding URL got 503 unavailable after I make public.

![image](https://user-images.githubusercontent.com/10163840/114553073-d572d700-9c97-11eb-80d5-a426fb299d25.png)

If possible, may I know when I can use a relative stable version of this feature? 
Appreciate if you let me know the timeline :)

**Reply by 2percentsilk** [reply]
Hi @dooriya,

Can you retry accessing the link?  We're seeing a few errors from folks where making it public the first time results in a 503 but refreshing 1-2 times causes it to work.
For reference, I've added your request ID to our investigation though and will report back as I have more details.

If you run into a 503 again, can you copy/paste the Request ID here as well?

**Reply by dooriya** [reply]
Thank you @2percentsilk!

Overall we got 2 issues when using the public port-forwarding endpoint:
1. Got **503 Service Unavailable** error when access the public endpoint, and usually retry doesn't work. (**Request ID: a63b990b2c5746a835a8e6d462717a1a**) 
    
    ![image](https://user-images.githubusercontent.com/10163840/114653579-ed3f6f00-9d1a-11eb-84d8-d4d140550a68.png)

2. The public endpoint still redirect for authentication even I "Make Public".
   
    ![image](https://user-images.githubusercontent.com/10163840/114653486-bff2c100-9d1a-11eb-90b8-c75fef754823.png)

**Reply by elitan** [reply]
I'm seeing the same issue. Refreshing multiple times does not fix the issue.

Request ID: 1e25b703ddff598fef104bae25b6371f

<img width="1082" alt="CleanShot 2022-06-12 at 08 16 39@2x" src="https://user-images.githubusercontent.com/331818/173219698-fc947933-f1e1-4eb5-9110-4af022e7e1f2.png">

**Comment by I-Dont-Remember** [comment]
❌ it appears that the GitHub proxy is mangling requests before they reach the Codespaces local server:

![image](https://user-images.githubusercontent.com/8552248/204400595-f213756f-3e1f-44d0-95a9-26b0a66318c2.png)

```
INFO:werkzeug:127.0.0.1 - - [28/Nov/2022 23:12:02] "½¹È°|ÎK­;kqdá-mWÊÇ6tÂ#íSÂòà~8À,À0©Ì¨ÌªÀ+À/$À(kÀ#À'gÀ" HTTPStatus.BAD_REQUEST -
ERROR:werkzeug:127.0.0.1 - - [28/Nov/2022 23:12:02] code 400, message Bad HTTP/0.9 request type ('\x16\x03\x01\x00½\x01\x00\x00¹\x03\x03*¿Îoè~r\x91°\x1b\x89+\x06\x10öõW\\×\x10æµA°ñ\x01')
INFO:werkzeug:127.0.0.1 - - [28/Nov/2022 23:12:02] "½¹*¿Îoè~r°+öõW\×æµA°ñ:dqP¢8À,À0©Ì¨ÌªÀ+À/$À(kÀ#À'gÀ" HTTPStatus.BAD_REQUEST -
```

### what i did
- Running a very simple Python Flask API and sending a POST request to it (hitting the `/health` endpoint which does nothing but respond with 200). Using port `4747` if that makes a difference - tried with both `https` and `http` and had same result.
- Switched to trying the `public endpoint` as suggested above to get past the authentication redirect, but then all the traffic reaching my server is garbled.
  -  I am fairly confident it's not my code, because it is working just fine with Ngrok and running on my actual computer at the same commit hash.
